[consensus] Use newtypes for Epoch, View and ViewDelta

andresilva opened the pull request:
Converts `Epoch` and `View` from type aliases to newtype structs wrapping `u64`, and adds `ViewDelta` for semantic distinction between points in time (`View`) and durations (`ViewDelta`). Implements varint encoding for efficient wire format.

Fixes #1723.

andresilva commented on consensus/src/marshal/mod.rs:
```diff
@@ -405,7 +405,7 @@ mod tests {
                     // Otherwise, finalize randomly.
                     if height == NUM_BLOCKS
-                        || utils::is_last_block_in_epoch(BLOCKS_PER_EPOCH, epoch).is_some()
+                        || utils::is_last_block_in_epoch(BLOCKS_PER_EPOCH, height).is_some()
```
This happened in a test and it's meaningless, but it's a bug that the new types caught!

patrick-ogrady commented on consensus/src/simplex/actors/voter/actor.rs:
```diff
@@ -648,7 +648,7 @@ impl<
 
         // Request proposal from application
-        debug!(view = self.view, "requested proposal from automaton");
+        debug!(view = %self.view, "requested proposal from automaton");
```
We use `?` elsewhere in the codebase? If this was auto-generated, add it to AGENTS.md

andresilva replied:
The reason I used display was to avoid this looking like `view=View(10)`. The display implementation just prints the bare number.

patrick-ogrady commented on consensus/src/simplex/actors/voter/actor.rs:
```diff
@@ -908,12 +905,17 @@ impl<
             }
-            cursor -= 1;
+
+            if let Some(previous) = cursor.previous() {
```
Can we `.expect()` here? The code before didn't have a problem assuming non-zero

andresilva commented:
> pondering if there is some way to reduce the number of .into()s in this PR?

Most of them come from feeding those values into configs in tests, I don't think those can be avoided. Will see about other cases.

andresilva commented on consensus/src/simplex/actors/voter/actor.rs:
```diff
@@ -998,7 +1000,7 @@ impl<
 
         // Update metrics
-        self.current_view.set(view as i64);
+        self.current_view.set(view.get() as i64);
```
In theory we could change the gauge type to `Gauge<View, Atomic<View>>` and implement `Atomic` for `View` by just using `AtomicU64`, but that's also quite ugly. I think it could make sense to create our own wrapper types that are more ergonomic to use, logged as #2151.

patrick-ogrady commented on consensus/src/types.rs:
```diff
@@ -1,73 +1,405 @@
+/// For convenience, type aliases [`EpochDelta`] and [`ViewDelta`] are provided and should
+/// be preferred in most code.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Delta<T>(u64, PhantomData<T>);
```
❤️

patrick-ogrady commented on consensus/src/types.rs:
```diff
@@ -1,73 +1,417 @@
+};
 
-/// Epoch is the type used to represent a distinct set of validators.
+/// Represents a distinct set of validators in the consensus protocol for a contiguous
```
nit: `Represents a distinct segment of a contiguous sequence of views.`

(we assign validators as a result but I don't think the raw epoch type needs to know that)

patrick-ogrady commented on consensus/src/types.rs:
```diff
@@ -1,73 +1,405 @@
+    fn from(value: u64) -> Self {
+        Self::new(value)
+    }
+}
```
nice

patrick-ogrady commented on consensus/src/types.rs:
```diff
@@ -76,15 +420,254 @@ mod tests {
     use super::*;
     use commonware_codec::{DecodeExt, Encode, EncodeSize};
 
+    // ===== Epoch Tests =====
```
nit: remove this

patrick-ogrady commented on consensus/src/types.rs:
```diff
@@ -76,15 +420,254 @@ mod tests {
+        }
+    }
+
+    // ===== Delta Tests =====
```
nit: remove this

patrick-ogrady commented on consensus/src/types.rs:
```diff
@@ -76,15 +420,254 @@ mod tests {
+        assert_eq!(ViewDelta::new(42), ViewDelta::new(42));
+    }
+
+    // ===== Round Tests =====
```
nit: remove this

patrick-ogrady commented on examples/reshare/src/dkg/actor.rs:
```diff
@@ -369,8 +372,8 @@ where
-                        if let Some(epoch) = next_epoch.checked_sub(2) {
-                            self.round_metadata.remove(&epoch.into());
+                        if let Some(epoch) = next_epoch.checked_sub(EpochDelta::new(2)) {
+                            self.round_metadata.remove(&U64::from(epoch.get()));
```
I think we could use `Epoch` directly here?

patrick-ogrady commented on examples/reshare/src/dkg/manager.rs:
```diff
@@ -171,7 +171,7 @@ where
 
         // If the node crashed in the middle of dealing, recover the dealer state from storage.
-        let dealer_meta = if let Some(meta) = store.get(&epoch.into()) {
+        let dealer_meta = if let Some(meta) = store.get(&U64::from(epoch.get())) {
```
I think we could support `into<U64>`?

patrick-ogrady reviewed (commented):
very close

BrendanChou commented on consensus/src/types.rs:
```diff
@@ -1,73 +1,400 @@
+    fn from(epoch: Epoch) -> Self {
+        U64::from(epoch.get())
+    }
+}
```
can you add to/from for u64 as well?

BrendanChou commented on consensus/src/types.rs:
```diff
@@ -1,73 +1,400 @@
+///
+/// Created by [`View::range`]. Iterates from start (inclusive) to end (exclusive).
+pub struct ViewRange {
+    current: View,
```
prefer `start` rather than `current` since it may be a "backwards-facing" range? Like the "current" might actually be at the end? I guess it doesn't matter much since it's private fields but would just be more consistent

Is there a reason this is better than just using [Range](https://doc.rust-lang.org/std/ops/struct.Range.html) over `View`?

BrendanChou commented on consensus/src/marshal/mod.rs:
```diff
@@ -740,13 +756,15 @@ mod tests {
             let proposal3 = Proposal {
-                round: Round::new(0, 3),
-                parent: 2,
+                round: Round::new(Epoch::new(0), View::new(3)),
```
worthwhile to have a `Round` method be able to take u64's directly? I see why you made it as safe as possible but this is almost too much boilerplate

BrendanChou commented on consensus/src/simplex/actors/voter/state.rs:
```diff
@@ -654,7 +654,8 @@ impl<E: Clock + Rng + CryptoRng + Metrics, S: Scheme, D: Digest> State<E, S, D>
     fn find_parent(&self, view: View) -> Result<(View, D), View> {
         // If the view is the genesis view, consider it to be its own parent.
-        let mut cursor = view.saturating_sub(1);
+        let mut cursor = view.previous().unwrap_or(View::zero());
```
nit: `GENESIS_VIEW` instead of `View::zero()`

andresilva commented on consensus/src/types.rs:
```diff
@@ -1,73 +1,400 @@
+///
+/// Round provides a total ordering across epoch boundaries, where rounds are
+/// ordered first by epoch, then by view within that epoch.
 #[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
```
`zero()` is still useful for `const fn`. Will add `Default`.

andresilva commented:
Gave this one final review myself, I think it should be good to go (once #2263 is merged). Only one unaddressed comment from @BrendanChou's review regarding `Round::new(Epoch::new(...), View::new(...))` vs `(Epoch::new(...), View::new(...)).into()`.

andresilva commented on consensus/src/simplex/actors/voter/state.rs:
```diff
@@ -682,9 +682,7 @@ impl<E: Clock + Rng + CryptoRng + Metrics, S: Scheme, D: Digest> State<E, S, D>
-        // FIXME
-        for cursor in (GENESIS_VIEW.get() + 1..=self.view.get()).rev() {
-            let cursor = View::new(cursor);
+        for cursor in View::range(GENESIS_VIEW.next(), self.view.next()).rev() {
```
`Step` is only available on nightly sadly :-(

patrick-ogrady commented on consensus/src/types.rs:
```diff
@@ -1,73 +1,388 @@
+    type Cfg = ();
+
+    fn read_cfg(buf: &mut impl Buf, _cfg: &Self::Cfg) -> Result<Self, Error> {
+        let value: u64 = UInt::read(buf)?.into();
```
conveniently, this isn't a breaking change ❤️
